### DISCUSSION TITLE
Integration with GeoArrow and Lonboard for large-scale interactive visualization/animation

### DISCUSSION THREAD

**Discussion by kylebarron** [xoolive/traffic#414] [Ideas]
👋 This is a really cool library, and I wonder if there might be some mutual benefit to integrating our work! 

In the quickstart you [note](https://traffic-viz.github.io/quickstart.html#making-maps)

> for quick interactive representations with **few elements**, the Leaflet widget is a good option

ipyleaflet indeed is not very performant. I've been working on an open source library called [Lonboard](https://github.com/developmentseed/lonboard) which connects to [deck.gl](https://deck.gl/). By using [Arrow](https://arrow.apache.org/) and Parquet efficiently, it can render ~3 million coordinates in an interactive map from Python in a few seconds. 

deck.gl was originally built by Uber and since they also wanted to visualize large numbers of moving objects, the JS library includes a `TripsLayer` that animates moving objects ([example](https://deck.gl/examples/trips-layer), note that reducing the "trail" slider on the right makes each trajectory clearer). 

I've started but haven't yet finished a [PR that connects this to Python](https://github.com/developmentseed/lonboard/pull/292). That PR's description has a screen recording of an example animation from ship AIS data, but I think air traffic data would make a really compelling demo. With that PR you should be able to render trajectories containing 3 million coordinates, and ideally be able to set a different color _per coordinate_ based on some attribute like current speed.

Under the hood, Lonboard works by converting data to the [GeoArrow data format](https://geoarrow.org/), which for an array of LineStrings is essentially raw arrays of coordinates for the entire column plus an offset array to determine at which indices each LineString starts and ends. So what I'd need to implement is the adapter from a `Traffic` object to a `pyarrow.Table` with timestamps at the same level of nesting as the LineString column.

Side note: I see in https://github.com/xoolive/rs1090 one of the goals is "exporting trajectory data to cross-platform formats such as JSON or parquet". You might be interested in using (Geo)Arrow for this, which would allow for sharing large data between Rust and Python without copies. I'm building [a Rust-Python (and Rust-Wasm) library for GeoArrow](https://github.com/geoarrow/geoarrow-rs) that benefits from being able to move Arrow data between Rust and Python for free.

Do you think this kind of integration makes sense? Is this interesting? If so, is there a data source to play with that contains multiple flights at the same time in the same airspace? I'm thinking something like a 30-minute window of flights in congested airspace like NYC would make for a really cool visualization. I haven't yet figured out a good way to access data like that though.

**Comment by xoolive** [comment]
Hello @kylebarron 

I am quite interested with your project, looks really cool!

So there's a reason why I am not using the geopandas from the start. That's a question I asked myself early on and the thing is that:
- most analyses we do are actually not related with the positional information,
- geopandas does not extend well when you want to attach more features beyond the lat/lon/alt data. 

So in traffic paradigm, a trajectory is not a line in a dataframe but a sub-dataframe.

I have some functions to iterate over flights, and the `summary()` maps flights to a row with attributes passed in parameters. So I tried the following code (which could be a start for what you need?) and it obviously needs some fine-tuning. :) This is data for Zurich airport which is already integrated in the library, but we can build a new dataset with New York at some point for your demo if you want. That's 2 months of data but you can reduce the scope easily.

Does it help?

Side note about Rust: I am interested too! For now, I am basically serializing vectors of JSON through pickle when passing decoded messages from Rust to Python, but it's not ideal... (I am new to Rust, this is my first project!)  Considering my comment above I am not sure it would fit well though, but will have a look. About what I read [here](https://github.com/geoarrow/geoarrow-rs/tree/main/python) wouldn't it be a smarter move to stick to a full Rust (more simple) implementation like https://github.com/busstoptaktik/geodesy/ ?

```python
# %%

import geopandas as gpd
from traffic.data.datasets import landing_zurich_2019

summary = (
    landing_zurich_2019.resample("10s")
    .summary(["start", "stop", "callsign", "icao24", "linestring"])
    .eval(desc="", max_workers=4)
)
gdf = gpd.GeoDataFrame(summary.rename(columns=dict(linestring="geometry")))

# %%
from lonboard import viz

viz(gdf)
```

![image](https://github.com/xoolive/traffic/assets/1360812/b6d9e343-e87b-4985-9c55-2a2d103f4297)

**Reply by kylebarron** [reply]
> So in traffic paradigm, a trajectory is not a line in a dataframe but a sub-dataframe.

Right I see in the intro doc that 

> [Flight](https://traffic-viz.github.io/api_reference/traffic.core.flight.html#traffic.core.Flight) is the core class offering representations, methods and attributes to single trajectories.
>
> [Traffic](https://traffic-viz.github.io/api_reference/traffic.core.traffic.html#traffic.core.Traffic) is the core class to represent collections of trajectories.

But I'm not trying to suggest that a Traffic instance should be just a GeoDataFrame. GeoArrow and GeoPandas are not the same, and in an Arrow representation, you can let each `Flight` be a logical row, even where some values in that row are themselves a list. But in Arrow this is no less performant than Numpy. 

Maybe an example with the Zurich data will be enlightening:

```py
from traffic.data.datasets import landing_zurich_2019
import numpy as np
import pyarrow as pa

static_col_names = [
    "callsign",
    "icao24",
    "lastseen",
    "flight_id",
    "runway",
    "origin",
    "initial_bearing",
    "initial_flow",
    "simple",
]
variable_col_names = [
    "timestamp",
    "altitude",
    "geoaltitude",
    "latitude",
    "longitude",
    "groundspeed",
    "onground",
    "track",
    "vertical_rate",
    "distance",
]

num_rows = landing_zurich_2019.data["flight_id"].nunique()

static_columns = {col_name: [] for col_name in static_col_names}
current_offset = 0
offsets = np.zeros(num_rows + 1, dtype=np.int32)

# NOTE: this approach assumes the input data is already sorted! I.e. that you have all
# of group 1 then all of group 2, etc.
for group_idx, (flight_id, group) in enumerate(
    landing_zurich_2019.data.groupby("flight_id")
):
    # Push a single scalar value for each static column
    for col_name in static_columns:
        static_columns[col_name].append(group[col_name].iloc[0])

    # Record the length of this group
    current_offset += len(group)
    offsets[group_idx + 1] = current_offset


pyarrow_static_columns = {
    col_name: pa.array(l) for col_name, l in static_columns.items()
}
pyarrow_nested_columns = {
    col_name: pa.ListArray.from_arrays(
        offsets, pa.array(landing_zurich_2019.data[col_name])
    )
    for col_name in variable_col_names
}

table = pa.table({**pyarrow_static_columns, **pyarrow_nested_columns})
```

So now this table has _nested_ columns:

```
pyarrow.Table
callsign: string
icao24: string
lastseen: timestamp[us, tz=UTC]
flight_id: string
runway: string
origin: string
initial_bearing: double
initial_flow: string
simple: bool
timestamp: list<item: timestamp[ns, tz=UTC]>
  child 0, item: timestamp[ns, tz=UTC]
altitude: list<item: double>
  child 0, item: double
geoaltitude: list<item: double>
  child 0, item: double
latitude: list<item: double>
  child 0, item: double
longitude: list<item: double>
  child 0, item: double
groundspeed: list<item: double>
  child 0, item: double
onground: list<item: bool>
  child 0, item: bool
track: list<item: double>
  child 0, item: double
vertical_rate: list<item: double>
  child 0, item: double
distance: list<item: double>
  child 0, item: double
```

But this is a really efficient representation because for all the static columns we only store values _once_ per row, instead of repeated for every row. So this uses 6.5x less memory than the pandas DataFrame:

```
landing_zurich_2019.data.memory_usage(deep=True).sum()
# 8550877073
table.nbytes
# 1322412459
```

With pandas v2, you can actually still represent this in pandas by using the Arrow dtypes

```
df_arrow = table.to_pandas(types_mapper=pd.ArrowDtype)
df_arrow.head(n=2)
```
<img width="727" alt="image" src="https://github.com/xoolive/traffic/assets/15164633/d16acd8a-a735-49d4-ba84-493e69c0e6a4">

And this arrow-based Pandas DataFrame still uses 6.5x less than the original Pandas DataFrame:
```
df_arrow.memory_usage(deep=True).sum()
# 1322412591
```

I think this little script above is most of what I need to test with traffic data. Lonboard uses this really efficient representation, and copies the nested `longitude` and `latitude` columns straight to the GPU.

**Reply by kylebarron** [reply]
https://github.com/xoolive/traffic/assets/15164633/b3126fe3-d789-4956-a981-7f7dad9b7a2f



```py

from lonboard import PathLayer, Map

# Create "GeoArrow" geometry column from our data
np_coords = np.column_stack([
    df['longitude'], df['latitude'], df['altitude']
]).ravel('C')
pa_coords = pa.FixedSizeListArray.from_arrays(np_coords, list_size=3)
pa_linestrings = pa.ListArray.from_arrays(offsets, pa_coords)
field = pa.field("geometry", pa_linestrings.type, metadata={'ARROW:extension:name': "geoarrow.linestring"})

# Here we select just a few columns to make the rendering faster, but you can change the `select` call
# Any attributes will be shown in the pop up when hovering
geo_table = table.append_column(field, pa_linestrings).select(['callsign', 'timestamp', 'altitude', 'geometry'])

layer = PathLayer(table=geo_table)
m = Map(layer)
m
```

**Reply by xoolive** [reply]
I see! Maybe I should look into converting from and to pyarrow tables then. I am not sure I can really get rid of the pandas things yet because:
1. I don't want to confuse users just yet
2. it is convenient to select pieces of trajectories with a nice operational semantic

Maybe, it would be nice to use this to store data anyway.

You say I can pass this transparently to/from Rust after that?

**Reply by kylebarron** [reply]
> I see! Maybe I should look into converting from and to pyarrow tables then. I am not sure I can really get rid of the pandas things yet because:
> 
> 1. I don't want to confuse users just yet
> 2. it is convenient to select pieces of trajectories with a nice operational semantic

💯 I'm not suggesting you change your API. This Arrow stuff is definitely advanced and in some early days. But it can be really efficient!

> You say I can pass this transparently to/from Rust after that?

Yeah! You can test with the `geoarrow-rust-core` PyPI package, though that only works with 2D geometries for now.

```py
# Create "GeoArrow" geometry column from our data
np_coords = np.column_stack([
    df['longitude'], df['latitude']
]).ravel('C')
pa_coords = pa.FixedSizeListArray.from_arrays(np_coords, list_size=2)
pa_linestrings = pa.ListArray.from_arrays(offsets, pa_coords)
field = pa.field("geometry", pa_linestrings.type, metadata={'ARROW:extension:name': "geoarrow.linestring"})
geo_table = table.append_column(field, pa_linestrings).select(['callsign', 'timestamp', 'altitude', 'geometry'])

from geoarrow.rust.core import GeoTable
%timeit rust_table = GeoTable.from_arrow(geo_table)
# 13.2 µs ± 425 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

it takes 13 microseconds to access this entire dataframe from rust! And no matter your data size, this step is always zero-copy! What's happening here is that the data access is now shared between the pyarrow library (written in C++) and the `geoarrow-rust` library (written in Rust). So now I have full access to the data in whatever rust-Python code I create.

**Reply by xoolive** [reply]
Thanks!

- If I take some time this week to add `Traffic.to_arrow_table()` and `Traffic.from_arrow_table()` would your code for the demo looks cleaner?
- Are you willing to discuss more fun visualisations we could implement with your library? (I have quite some ideas actually, with no deadline of course… would be nice if some of them meet your interests?)
- I think I will start digging into geoarrow-rs from the rs1090 side (so first: full Rust) and probably later look into this aspect to improve some trajectory processing on the Rust side (from Python).
- I actually had some of my many opened tabs on your project there for few weeks 😅 Thinking of doing more things here: https://observablehq.com/collection/@openaviation/open-aviation happy to do anything with four hands to showcase your library if you want

**Reply by xoolive** [reply]
Hi @kylebarron could you edit the snippet for the video to provide `offsets` and `table` definitions please?  
(now working on item 1)

Edit: just realised it was above 🤦‍♂️

**Reply by kylebarron** [reply]
> Edit: just realised it was above 🤦‍♂️

Yeah, sorry if it wasn't clear; each code segment built on the previous one. Note that the above code uses a simplifying assumption that the data is already correctly sorted, first by flight segment and then by time, which we might want to verify on the input data.

I'm happy to discuss anything around GeoArrow! Either on the Rust, Python, or JS sides. The Rust API of geoarrow-rs isn't documented as well as the Python side because I expect more of the initial users to be Python users. 

Any of these visualizations in Python in Lonboard are also possible in pure JS using the underlying [`@geoarrow/deck.gl-layers`](https://github.com/geoarrow/deck.gl-layers) library, so you could use that from an observable notebook. Though there are some known JS bundling bugs.

**Reply by xoolive** [reply]
Hi @kylebarron could I get errors reproducing your example, probably because I couldn't get the CRS set. I am not sure how to set it without going through the geodataframe you don't build the way you constructed things.

**Reply by kylebarron** [reply]
Without the CRS you'll get a warning, but it'll still render (you can see that in your screenshot above). 

Can you say anything more about what's not working for you?

**Reply by xoolive** [reply]
I get the widget, the warning, and no trajectory data in the widget. The only use case I get working is my screenshot above going through GeoPandas.

**Reply by xoolive** [reply]
@kylebarron I can't share my notebook here, will send it on your gmail.

**Reply by xoolive** [reply]
Tried again, after upgrading the environment.
Looks more stable, but still got an empty widget :(

<img width="510" alt="Capture d’écran 2024-04-02 à 22 57 57" src="https://github.com/xoolive/traffic/assets/1360812/822456b0-c306-4639-8cfa-fb5695b760f2">

Still working with the same data through geopandas first

**Reply by kylebarron** [reply]
> Still working with the same data through geopandas first

If it's working through geopandas, then it probably means the geopandas -> geoarrow conversion code I wrote isn't stable enough.

Maybe try sorting the table by flight id then timestamp, and then running my above code. The code was predicated on a precise ordering within the table.

**Reply by kylebarron** [reply]
> Looks more stable, but still got an empty widget :(

You can also use something like `width_min_pixels=10` to make sure that you aren't just missing data that's rendered too small to see.

**Reply by xoolive** [reply]
@kylebarron Finally managed something smaller scale, starting from adding the airport layout helped debugging.

![image](https://github.com/xoolive/traffic/assets/1360812/972dd698-bf07-4c33-a62b-ad2609f05b71)

weird though, it feels more heavy than plotly for instance... maybe there's something I do wrong

**Reply by xoolive** [reply]
Just some thinking aloud:
- the plateau in altitude after take-off in order to take speed before climbing further looks cool!
- when aircraft is on the ground and not moving (much), it appears as spikes (i guess it's something related to the small movements related to the `width_min_pixel` parameter), I should maybe run some Douglas Peucker first;
- when creating the pyarrow, does the geoarrow-rs have a role to play somewhere help for speed/feeling smooth, or is it a different story?
- or maybe the 3D makes it just heavier... but still...

**Reply by kylebarron** [reply]
> it feels more heavy than plotly for instance

By "heavy" are you referring to the width of the lines when plotted? Or something else? You can customize the width of lines by setting `width_min_pixels` on the layer. You might want to reduce it for your case.

> when creating the pyarrow, does the geoarrow-rs have a role to play somewhere help for speed/feeling smooth, or is it a different story?

Ideally geoarrow-rs in the future will help with operations in memory to work with the data. But today it doesn't yet support 3D coordinates, so it can't be used with this

**Reply by xoolive** [reply]
@kylebarron I mean there is some lag when I pan/zoom on the map

**Reply by kylebarron** [reply]
How many coordinates are you visualizing?

**Reply by xoolive** [reply]
Not so many, have to check for the airport but the flight is like 6000...

**Reply by kylebarron** [reply]
If the map is laggy with 6000 coordinates, then there's something wrong at the deck layer. Like, can you ensure that GPU hardware acceleration is turned on? Often being slow (especially on Linux) can be from software emulating GPU operations

**Reply by xoolive** [reply]
That could very well be indeed. Do you have a reference explaining how to check that?

**Reply by kylebarron** [reply]
It should be in your browser settings. But it's different for each browser, so I don't have a foolproof suggestion

**Reply by xoolive** [reply]
I just quickly tried from home with my old x86_64 Mac:
- in vscode it's similar... not great...;
- in the browser, it's super smooth (just not so convenient with the trackpad to rotate the view and get the option menu at the same time 🤣 ).

Will check both options from work on Linux on Monday. (You already know the option in vscode maybe? I am never starting a jupyter server actually... 😰)

**Reply by kylebarron** [reply]
> in the browser, it's super smooth

Ok, so at least we've confirmed that in _some_ setup it works well. Hopefully that means the issue is outside of my GeoArrow glue and instead in some browser webview interactions. 

For me, on a Mac M2, it's just as fast in VSCode as it is in Chrome. 

> You already know the option in vscode maybe? I am never starting a jupyter server actually... 😰

Vscode will automatically start a jupyter process under the hood for you

**Reply by xoolive** [reply]
yes it's probably on me, but if it's good I get through everything before others do, at least I can process most of the stuff under the hood then (that's how `traffic` was born)

> vscode

I usually use the `# %%` nodes to run an ipykernel. I would like very much that it works as well on my Linux too!

**Reply by kylebarron** [reply]
> I usually use the `# %%` nodes to run an ipykernel.

You don't actually have to define those blocks manually. See the gif in https://github.com/kylebarron/vscode-jupyter-python

**Reply by xoolive** [reply]
🤯

**Reply by xoolive** [reply]
@kylebarron so in Ubuntu, vscode through snap doesn't get access to gpu... but with the .deb file it does... mystery solved.

now i can start wrapping things :+1:
